FLEXIBLE_INDEXES_NOTES.md

```diff
@@ -0,0 +1,396 @@
+- plotting
+    - `plot.pcolormesh` "infers" interval breaks along axes, which are really inferred `bounds` for the appropriate indexes.
+    - `plot.step` again uses `bounds`. In fact, we may even want `step` to be the default 1D plotting function if the axis has `bounds` attached.
+
+It would be reasonable to first restrict those methods to the indexes that are currently available in Xarray, and maybe extend the `XarrayIndex` API later upon request when the opportunity arises.
+
+Conversely, nothing should prevent implementing "non-standard" API in 3rd-party `XarrayIndex` subclasses that could be used in DataArray/Dataset extensions (accessors). For example, we might want to reuse a `KDTree` index to compute k-nearest neighbors (returning a DataArray/Dataset with a new dimension) and/or the distances to the nearest neighbors (returning a DataArray/Dataset with a new data variable).
+
+### 2.8 Index encoding
+
+Indexes don't need to be directly serializable since we could (re)build them from their corresponding coordinate(s). However, we may take advantage that some indexes could be encoded/decoded to/from a set of arrays that would allow optimized reconstruction and/or storage, e.g.,
+
+- `pandas.MultiIndex` -> `index.levels` and `index.codes`
+- Scikit-learn's `KDTree` and `BallTree` that use an array-based representation of an immutable tree structure
+
+## 3. Index representation in DataArray/Dataset's `repr`
+
+Since indexes would become 1st class citizen of Xarray's data model, they deserve their own section in Dataset/DataArray `repr` that could look like:
+
+```
+<xarray.DataArray (x: 2, y: 2)>
+array([[5.4, 7.8],
+       [6.2, 4.7]])
+Coordinates:
+  * lon      (x, y) float64 10.2 15.2 12.6 17.6
+  * lat      (x, y) float64 40.2 45.6 42.2 47.6
+  * x        (x) float64 200.0 400.0
+  * y        (y) float64 800.0 1e+03
+Indexes:
+  lat, lon     <SpatialIndex coords=(lat, lon) dims=(x, y)>
+  x            <PandasIndexWrapper>
+  y            <PandasIndexWrapper>
+```
+
+To keep the `repr` compact, we could:
+
+- consolidate entries that map to the same index object, and have an short inline repr for `XarrayIndex` object
+- collapse the index section by default in the HTML `repr`
+- maybe omit all trivial indexes for 1D coordinates that match the dimension name
+
+## 4. `IndexVariable`
+
+`IndexVariable` is currently used to wrap an `pandas.Index` as a variable. Although the main goal of this refactoring is to decouple indexes and variables (coordinates), `IndexVariable` could still be useful to:
+
+- ensure that all coordinates with an index are immutable (see [Section 2.4.1](#241-mutable-coordinates))
+  - do not set values directly, do not (re)chunk (even though it may be already chunked), do not load, do not convert to sparse/dense, etc.
+- directly reuse index's data when that's possible
+  - in the case of a `pandas.Index`, it makes little sense to have duplicate data (e.g., as a NumPy array) for its corresponding coordinate
```

Comment:
In the long term, I think it would probably be ideal to eliminate `IndexVariable`, which is just one more thing that can get out of sync when setting/resetting a index.

Instead, we could ensure that the `data` associated with a `Variable` is immutable by writing a custom array wrapper class like `CopyOnWriteArray` that raises an error from `__setitem__`.